File: src/symnmf.py
import argparse
import os
import numpy as np
import symnmf_c  # Import C module

GOALS = ['symnmf', 'sym', 'ddg', 'norm']
# Set random seed at the beginning of the code
np.random.seed(1234)

def parse_arguments():
    """Parse command line arguments"""
    parser = argparse.ArgumentParser(description="Run symnmf using C extension")
    parser.add_argument('k', type=int, help='Number of clusters')
    parser.add_argument('goal', type=str, choices=GOALS, help='Goal')
    parser.add_argument('file_name', type=str, help='Input file path')
    return parser.parse_args()

def read_data_points(file_name):
    """Read data points from file and convert to numpy array"""
    points = []
    with open(file_name, 'r') as f:
        lines = f.readlines()
        for line in lines:
            # Convert line to list of floats
            row = [float(x) for x in line.strip().replace(',', ' ').split() if x]
            if row:
                points.append(row)
    return np.array(points)

def initialize_H(W, k):
    m = np.mean(W)
    n = W.shape[0]
    upper = 2.0 * np.sqrt(m / k)
    H = np.random.uniform(0.0, upper, size=(n, k))
    return H

def handle_goal(goal, X, k):
    """Call appropriate C function based on goal"""
    if goal == 'symnmf':
        W = symnmf_c.norm(X)
        H = initialize_H(W, k)
        return symnmf_c.symnmf(H, W)
    elif goal == 'sym':
        return symnmf_c.sym(X)
    elif goal == 'ddg':
        return symnmf_c.ddg(X)
    elif goal == 'norm':
        return symnmf_c.norm(X)
    else:
        raise ValueError("Invalid goal")

def print_matrix(result):
    """Output the required matrix separated by comma, each row in a line"""
    for row in result:
        print(','.join(f'{val:.4f}' for val in row))

def main():
    """Main execution function"""
    try:
        args = parse_arguments()
        assert args.goal in ['symnmf', 'sym', 'ddg', 'norm'], "Invalid goal"
        assert os.path.exists(args.file_name), "File does not exist"
        X = read_data_points(args.file_name)
        if args.goal == 'symnmf':
            assert 1 <= args.k < X.size, "k must be greater than 1 and less than number of data points"
        result = handle_goal(args.goal, X, args.k)
        print_matrix(result)
    except Exception as e:
        print(f"An Error Has Occurred: {e}")
        exit(1)

if __name__ == "__main__":
    main()

File: src/symnmf.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "symnmf.h"

/* Helper function to print error and cleanup */
static void error_exit(double **points, int n) {
    printf("An Error Has Occurred\n");
    cleanup_points(points, n);
    exit(1);
}

/* Process sym goal */
static void process_sym(double **points, int n, int dim) {
    double **A = alloc_matrix(n, n);
    if (!A) error_exit(points, n);
    
    build_similarity_matrix(points, n, dim, A);
    print_matrix(A, n, n);
    free_matrix(A);
}

/* Process ddg goal */
static void process_ddg(double **points, int n, int dim) {
    double **A = alloc_matrix(n, n);
    double **D;
    if (!A) error_exit(points, n);
    
    build_similarity_matrix(points, n, dim, A);
    
    D = alloc_matrix(n, n);
    if (!D) {
        free_matrix(A);
        error_exit(points, n);
    }
    
    build_degree_matrix(A, n, D);
    print_matrix(D, n, n);
    
    free_matrix(A);
    free_matrix(D);
}

/* Process norm goal */
static void process_norm(double **points, int n, int dim) {
    double **A = alloc_matrix(n, n);
    double **D;
    double **W;
    if (!A) error_exit(points, n);
    
    build_similarity_matrix(points, n, dim, A);
    
    D = alloc_matrix(n, n);
    if (!D) {
        free_matrix(A);
        error_exit(points, n);
    }
    
    build_degree_matrix(A, n, D);
    
    W = alloc_matrix(n, n);
    if (!W) {
        free_matrix(A);
        free_matrix(D);
        error_exit(points, n);
    }
    
    build_normalized_similarity_matrix(A, D, n, W);
    print_matrix(W, n, n);
    
    free_matrix(A);
    free_matrix(D);
    free_matrix(W);
}

int main(int argc, char *argv[]) {
    char *goal;
    char *file_name;
    double **points = NULL;
    int n = 0, dim = 0;
    
    if (argc != 3) {
        printf("An Error Has Occurred\n");
        return 1;
    }
    
    goal = argv[1];
    file_name = argv[2];
    
    /* Read input points */
    if (!read_points(file_name, &points, &n, &dim)) {
        printf("An Error Has Occurred\n");
        return 1;
    }
    
    /* Process based on goal */
    if (strcmp(goal, "sym") == 0) {
        process_sym(points, n, dim);
    } else if (strcmp(goal, "ddg") == 0) {
        process_ddg(points, n, dim);
    } else if (strcmp(goal, "norm") == 0) {
        process_norm(points, n, dim);
    } else {
        printf("An Error Has Occurred\n");
        cleanup_points(points, n);
        return 1;
    }
    
    cleanup_points(points, n);
    return 0;
}
